From e7a4097c42c46a96a7765b4342cf0979025a7d5f Mon Sep 17 00:00:00 2001
From: Marcin Wojtas <mw@semihalf.com>
Date: Thu, 21 Nov 2019 07:21:25 +0100
Subject: [PATCH 0998/1239] spi: mvebu_a3700_spi: prevent configuring speed
 exceeding max controller freq

This patch adds a limitation in the mvebu_a3700_spi driver
set_speed hook, which prevents setting too high transfer
speed.

Change-Id: Icef1778fca1685545100b05bb364c9fec9d0ff32
Signed-off-by: Marcin Wojtas <mw@semihalf.com>
Reviewed-on: https://sj1git1.cavium.com/19040
Reviewed-by: Kostya Porotchkin <kostap@marvell.com>
Tested-by: Kostya Porotchkin <kostap@marvell.com>
---
 drivers/spi/mvebu_a3700_spi.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/spi/mvebu_a3700_spi.c b/drivers/spi/mvebu_a3700_spi.c
index 1b6db6064d..58ed409fd5 100644
--- a/drivers/spi/mvebu_a3700_spi.c
+++ b/drivers/spi/mvebu_a3700_spi.c
@@ -383,9 +383,16 @@ static int mvebu_spi_xfer(struct udevice *dev, unsigned int bitlen,
 static int mvebu_spi_set_speed(struct udevice *bus, uint hz)
 {
 	struct mvebu_spi_platdata *plat = dev_get_platdata(bus);
+	struct dm_spi_bus *spi = dev_get_uclass_priv(bus);
 	struct spi_reg *reg = plat->spireg;
 	u32 data, prescale;
 
+	if (hz > spi->max_hz) {
+		debug("%s: limit speed to the max_hz of the bus %d\n",
+		      __func__, spi->max_hz);
+		hz = spi->max_hz;
+	}
+
 	data = readl(&reg->cfg);
 
 	prescale = DIV_ROUND_UP(clk_get_rate(&plat->clk), hz);
-- 
2.29.0

